src/ore/src/collections.rs

```diff
@@ -65,3 +66,38 @@ where
         }
     }
 }
+
+/// Extension methods for associative collections.
+pub trait AssociativeExt<K, V> {
+    /// Inserts a key and value, panicking with
+    /// a given message if a true
+    /// insert (as opposed to an update) cannot be done
+    /// because the key already existed in the collection.
+    fn expect_insert(&mut self, k: K, v: V, msg: &str);
+    /// Inserts a key and value, panicking if a true
+    /// insert (as opposed to an update) cannot be done
+    /// because the key already existed in the collection.
+    fn unwrap_insert(&mut self, k: K, v: V) {
+        self.expect_insert(k, v, "called `unwrap_insert` for an already-existing key")
+    }
+}
+
+impl<K, V> AssociativeExt<K, V> for HashMap<K, V>
+where
+    K: Eq + std::hash::Hash,
+{
+    fn expect_insert(&mut self, k: K, v: V, msg: &str) {
+        let old = self.insert(k, v);
+        assert!(old.is_none(), "{}", msg)
```

Comment:
maybe include the old value in the print? I think it'd be pretty informative -- even if it means you have to require that `K: Debug`?

Reply:
That can't be done in a zero-cost way -- you either have to clone the value when inserting, or do an extra lookup.

Reply:
(AFAICT)

Reply:
I was suggesting something like `assert!(old.is_none(), "{}: found {:?}", msg, old)` that shouldn't have extra work unless you're about to panic in which case 🤷 

but if you think requiring `K: Debug` makes this less useful, keeping as-is is fine by me

Reply:
Oh, I see. Yeah, if you could print the key and both the old and new values it'd maybe be useful enough to be outweigh the downside of requiring `Debug`, but I kinda feel like _just_ printing the old value doesn't.

Reply:
You can make it zero cost with the entry API.